### PR TITLE
Implement dual ESM + CJS compatibility in style-spec

### PR DESCRIPTION
**Pull request by rreusser**
This PR proposes three changes to style-spec to allow ES module + CommonJS compatibility while avoiding a major version bump or dropping CommonJS compatibility.

1. Rename the comnonjs dist file from `dist/index.js` to `dist/index.cjs`. This is encapsulated by the entry point and hints to `node` that it is a commonjs implementation. If it remains `index.js`, then `node` will fail and interpret it as a module.
2. Add an `"exports": {"require": , "import": }` field to `package.json`. This instructs node and bundlers how to import either one or the other, depending on whether it wants a commonjs or es module implementation.
3. Finally, add `"import": "./"` to the `"exports"` field so that ES module uses may import files directly. (Additionally they'll also need to be flow-aware, but this pattern is in current usage, is unchanged, and is a matter for the developer to worry about.)

cc @arindam1993 @tristen @elifitch

### CODE REVIEW

File: src/style-spec/package.json [hunk 1]
```diff
@@ -9,14 +9,22 @@
     "mapbox-gl-js"
   ],
   "license": "ISC",
-  "main": "./dist/index.js",
-  "module": "./dist/index.es.js",
+  "main": "./dist/index.es.js",
```
Comment (rreusser) [inline src/style-spec/package.json:12]: I don't understand the interaction of this with `"type": "module"`. Does this field even matter, or does the `"."` field in `"exports"` override this with environment-specific values?

Comment (rreusser) [reply src/style-spec/package.json]: This affects browserify usage. I've reverted this to its original state and confirmed that browserify, node+cjs and node+esm all work with this setup. Continuing on and checking more setups…

File: src/style-spec/package.json [hunk 2]
```diff
@@ -9,14 +9,23 @@
...
+  "main": "./dist/index.cjs",
   "module": "./dist/index.es.js",
   "type": "module",
+  "exports": {
+    ".": {
+      "require": "./dist/index.cjs",
+      "import": "./dist/index.es.js"
+    },
+    "./": {
+      "import": "./"
```
Comment (rreusser) [inline src/style-spec/package.json:21]: Note: Unlike esm, the presence of `"exports"` apparently does not affect which files `require` is allowed to require, so I've not explicitly exported files for require (also, I'm not aware this pattern is even in usage, based on a [search of the mapbox org](https://github.com/search?l=JavaScript&p=1&q=org%3Amapbox+mapbox-gl-style-spec&type=Code))

**Comment by rreusser** [comment]
For posterity, making a note of the specific error message that led to the `.cjs` extension. After specifying `require` vs. `import` exports, this code run with node@14:

```js
const spec = require('style-spec');
```

and which is consistent with usage in the mapbox org, yields this error:

```
internal/modules/cjs/loader.js:1080
      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
      ^
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/dist/index.js
require() of ES modules is not supported.
require() of /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/dist/index.js from /Users/rickyreusser/mapbox/gl-js-release/src/test.cjs is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/package.json.
```

It presents three solutions:

> rename index.js to end in .cjs

This is what I've done.

> change the requiring code to use import()

This is a lot more work, as it puts the burden on uses at mapbox and in the wild to track this down and update.

> or remove "type": "module" 

This is a possibility. I still haven't ruled it out, though it seems more likely to cause confusion and seems less future-aware since we maybe prefer to be forward-thinking and consider this first an ES module, and second with CommonJS compatibility which may be removed in the future.

**Review by karimnaaji — APPROVED** [review]
Thanks for the thorough details and bringing me up to speed on this issue. To my knowledge, this seems safe considering the low number of dependents there are (thanks for checking internal references within the org) and fixes the upgrade issue. Some minor concerns I had concerning this:
- Whether renaming to `.cjs` suggested here is a recommended practice, and it seems like it is directly mentioned in error reporting from node 14:
```
internal/modules/cjs/loader.js:1080
      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
      ^
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/dist/index.js
require() of ES modules is not supported.
require() of /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/dist/index.js from /Users/rickyreusser/mapbox/gl-js-release/src/test.cjs is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/rickyreusser/mapbox/gl-js-release/mapbox-gl-js/src/style-spec/package.json.
```
- If we dropped cjs support for any reason, we would simply remove the "exports" entirely.

**Pull request merged**